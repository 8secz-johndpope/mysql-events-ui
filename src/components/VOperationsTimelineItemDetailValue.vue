<script>
export default {
  name: 'OperationsTimelineItemDetailValue',
  props: {
    prefix: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    value: {
      required: true,
    },
    highlight: {
      type: Boolean,
      default: false,
    },
    add: {
      type: Boolean,
      default: false,
    },
    remove: {
      type: Boolean,
      default: false,
    }
  }
}
</script>


<template>
  <div
    :class="{
      'value--add': add,
      'value--remove': remove,
    }"
    class="value"
  >
    {{ prefix }}
    <span :class="{
      'highlight': highlight
    }">
      <strong>{{ name }}:</strong>
      {{ value }}
    </span>
  </div>
</template>

<style scoped>
.value {
  padding-left: 10px;
  padding-top: 5px;
  padding-bottom: 5px;
  text-align: left;
}

.value--remove {
  background-color: #ffeef0;
}

.value--add {
  background-color: #e6ffed;
}

.highlight {
  background-color: #acf2bd;
}
</style>
