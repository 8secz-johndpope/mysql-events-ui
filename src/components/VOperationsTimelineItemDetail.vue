<script>
import VOperationsTimelineItemDetailValue from '@/components/VOperationsTimelineItemDetailValue'

export default {
  props: {
    type: {
      type: String,
      default: '',
    },
    rows: {
      type: Array,
      default: () => [],
    }
  },
  components: {
    VOperationsTimelineItemDetailValue,
  },
  computed: {
    showBefore() {
      return (this.rows.length && this.rows[0].before)
    },
    showAfter() {
      return (this.rows.length && this.rows[0].after)
    },
    before() {
      if (!this.rows.length || !this.rows[0].before) return true
      const rows = this.rows[0].before
      return Object
        .keys(rows)
        .map(key => ({
          prefix: '-',
          name: key,
          value: rows[key],
        }))
    },
    after() {
      if (!this.rows.length || !this.rows[0].after) return true
      const rows = this.rows[0].after
      const before = this.rows[0].before
      return Object
        .keys(rows)
        .map(key => ({
          prefix: '+',
          name: key,
          value: rows[key],
          highlight: before && (before[key] !== rows[key])
        }))
    }
  }
}
</script>

<template>
  <div class="timeline-item-detail">
    <div class="timeline-item-detail__content">
      <div v-if="showBefore">
        <VOperationsTimelineItemDetailValue
          v-for="row in before"
          :key="row.value"
          :prefix="row.prefix"
          :name="row.name"
          :value="row.value"
          :highlight="row.highlight"
          remove
        />
      </div>

      <div v-if="showAfter">
        <VOperationsTimelineItemDetailValue
          v-for="row in after"
          :key="row.value"
          :prefix="row.prefix"
          :name="row.name"
          :value="row.value"
          :highlight="row.highlight"
          add
        />
      </div>
    </div>
  </div>
</template>

<style scoped>
.timeline-item-detail {
  position: absolute;
  top: 100%;
  left: 0;
  width: 100%;
	animation: fadeIn 0.3s ease-in both;
}

@keyframes fadeIn {
	from {
		opacity: 0;
		transform: translate3d(0, -20%, 0);
	}
	to {
		opacity: 1;
		transform: translate3d(0, 0, 0);
	}
}
</style>
