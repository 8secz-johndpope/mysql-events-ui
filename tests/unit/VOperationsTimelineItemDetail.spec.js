import { shallowMount } from '@vue/test-utils'
import VOperationsTimelineItemDetail from '@/components/VOperationsTimelineItemDetail'
import VOperationsTimelineItemDetailValue from '@/components/VOperationsTimelineItemDetailValue'
import socketFixture from './fixtures/socketServer'

describe('VOperationsTimelineItem', () => {
  let props

  const build = () => {
    const wrapper = shallowMount(VOperationsTimelineItemDetail, {
      propsData: props,
    })

    return {
      wrapper,
      content: () => wrapper.find('.timeline-item-detail__content'),
      values: () => wrapper.findAll(VOperationsTimelineItemDetailValue),
    }
  }

  beforeEach(() => {
    props = {
      type: '',
      rows: [],
    }
  })

  it('renders the component correctly', () => {
    // arranje
    const { wrapper } = build()

    // assert
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders correct INSERT operation details', () => {
    // arranje
    props.type = socketFixture.response.INSERT.type
    props.rows = socketFixture.response.INSERT.affectedRows
    const { values } = build()
    const firstValue = values().at(0)

    // assert
    expect(firstValue.props().prefix).toBe('+')
    expect(firstValue.props().name).toBe('id')
    expect(firstValue.props().value).toBe(4)
    expect(firstValue.props().highlight).toBe(false)
    expect(firstValue.props().add).toBe(true)
  })

  it('renders correct UPDATE operation details', () => {
    // arranje
    props.type = socketFixture.response.UPDATE.type
    props.rows = socketFixture.response.UPDATE.affectedRows
    const { values } = build()
    const firstValue = values().at(0)
    const secondValue = values().at(1)

    // assert
    expect(firstValue.props().prefix).toBe('-')
    expect(firstValue.props().name).toBe('id')
    expect(firstValue.props().value).toBe(4)
    expect(firstValue.props().highlight).toBe(false)
    expect(firstValue.props().remove).toBe(true)

    expect(secondValue.props().prefix).toBe('+')
    expect(secondValue.props().name).toBe('id')
    expect(secondValue.props().value).toBe(5)
    expect(secondValue.props().highlight).toBe(true)
    expect(secondValue.props().add).toBe(true)
  })

  it('renders correct DELETE operation details', () => {
    // arranje
    props.type = socketFixture.response.DELETE.type
    props.rows = socketFixture.response.DELETE.affectedRows
    const { values } = build()
    const firstValue = values().at(0)

    // assert
    expect(firstValue.props().prefix).toBe('-')
    expect(firstValue.props().name).toBe('id')
    expect(firstValue.props().value).toBe(5)
    expect(firstValue.props().highlight).toBe(false)
    expect(firstValue.props().remove).toBe(true)
  })
})