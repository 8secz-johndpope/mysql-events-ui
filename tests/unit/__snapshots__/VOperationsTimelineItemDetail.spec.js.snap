// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VOperationsTimelineItem renders the component correctly 1`] = `
<div class="timeline-item-detail">
  <div class="timeline-item-detail__content">
    <!---->
    <!---->
  </div>
</div>
`;
